File: generate-blog-list.py
#!/usr/bin/env python3
"""
Blog Post List Generator (Python Version)
Scans the /blog folder and creates blog-posts.json with metadata from each HTML file.
"""

import os
import json
import re
from datetime import datetime
from pathlib import Path

# Configuration for blog posts
BLOG_CONFIG = {
    # Map of filename to emoji icon
    'icons': {
        'boost-wifi-speed.html': '📡',
        'wifi-security-wpa3-vs-wpa2.html': '🔒',
        'router-placement-guide.html': '🏠',
        'understanding-speed-tests.html': '⚡',
        'gaming-wifi-latency.html': '🎮',
        'wifi6-vs-wifi6e.html': '🌐',
        'secure-home-wifi.html': '👥',
        'mesh-wifi-vs-routers.html': '📶',
        'troubleshooting-wifi.html': '🛠️',
        'guest-network-setup.html': '💼'
    },
    # Default icon if not specified
    'default_icon': '📝'
}

def extract_metadata(html_content, filename):
    """Extract metadata from HTML file"""
    
    # Extract title
    title_match = re.search(r'<title>(.*?)</title>', html_content, re.IGNORECASE)
    title = title_match.group(1).split('|')[0].strip() if title_match else filename.replace('.html', '')
    
    # Extract description
    desc_match = re.search(r'<meta name="description" content="(.*?)"', html_content, re.IGNORECASE)
    description = desc_match.group(1) if desc_match else ''
    
    # Extract keywords
    keywords_match = re.search(r'<meta name="keywords" content="(.*?)"', html_content, re.IGNORECASE)
    keywords = [kw.strip() for kw in keywords_match.group(1).split(',')] if keywords_match else []
    
    return {
        'title': title,
        'description': description,
        'keywords': keywords
    }

def generate_tags(keywords):
    """Generate tags from keywords (take first 2-3)"""
    return [kw.title() for kw in keywords[:2]]

def estimate_reading_time(html_content):
    """Estimate reading time based on content length"""
    # Remove HTML tags
    text = re.sub(r'<[^>]*>', ' ', html_content)
    # Count words
    words = [w for w in text.split() if len(w) > 0]
    word_count = len(words)
    # Average reading speed: 200 words per minute
    minutes = max(1, round(word_count / 200))
    return f"{minutes} min read"

def get_file_date(filepath):
    """Get file creation/modification date"""
    timestamp = os.path.getmtime(filepath)
    date = datetime.fromtimestamp(timestamp)
    return date.strftime('%b %d, %Y')

def generate_blog_list():
    """Main function to generate blog list"""
    script_dir = Path(__file__).parent
    blog_dir = script_dir / 'blog'
    
    if not blog_dir.exists():
        print('❌ Blog directory not found!')
        return
    
    # Get all HTML files
    blog_files = list(blog_dir.glob('*.html'))
    
    blog_posts = []
    for filepath in blog_files:
        filename = filepath.name
        
        try:
            with open(filepath, 'r', encoding='utf-8') as f:
                html_content = f.read()
            
            # Extract metadata
            metadata = extract_metadata(html_content, filename)
            tags = generate_tags(metadata['keywords'])
            reading_time = estimate_reading_time(html_content)
            date = get_file_date(filepath)
            icon = BLOG_CONFIG['icons'].get(filename, BLOG_CONFIG['default_icon'])
            
            blog_post = {
                'filename': filename,
                'url': f'blog/{filename}',
                'title': metadata['title'],
                'description': metadata['description'],
                'tags': tags,
                'readingTime': reading_time,
                'date': date,
                'icon': icon,
                'dateObj': datetime.strptime(date, '%b %d, %Y').timestamp()  # For sorting
            }
            
            blog_posts.append(blog_post)
            
        except Exception as e:
            print(f'⚠️  Error processing {filename}: {str(e)}')
            continue
    
    # Sort by date (newest first)
    blog_posts.sort(key=lambda x: x['dateObj'], reverse=True)
    
    # Remove dateObj before saving (used only for sorting)
    for post in blog_posts:
        del post['dateObj']
    
    # Save to JSON file
    output_path = script_dir / 'blog-posts.json'
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(blog_posts, f, indent=2, ensure_ascii=False)
    
    print(f'✅ Generated blog list with {len(blog_posts)} posts')
    print(f'📝 Saved to: {output_path}')

if __name__ == '__main__':
    generate_blog_list()

File: generate-blog-list.js
const fs = require('fs');
const path = require('path');

// Configuration for blog posts
const blogConfig = {
    // Map of filename to emoji icon
    icons: {
        'boost-wifi-speed.html': '📡',
        'wifi-security-wpa3-vs-wpa2.html': '🔒',
        'router-placement-guide.html': '🏠',
        'understanding-speed-tests.html': '⚡',
        'gaming-wifi-latency.html': '🎮',
        'wifi6-vs-wifi6e.html': '🌐',
        'secure-home-wifi.html': '👥',
        'mesh-wifi-vs-routers.html': '📶',
        'troubleshooting-wifi.html': '🛠️',
        'guest-network-setup.html': '💼'
    },
    // Default icon if not specified
    defaultIcon: '📝'
};

// Extract metadata from HTML file
function extractMetadata(htmlContent, filename) {
    const titleMatch = htmlContent.match(/<title>(.*?)<\/title>/);
    const descMatch = htmlContent.match(/<meta name="description" content="(.*?)"/);
    const keywordsMatch = htmlContent.match(/<meta name="keywords" content="(.*?)"/);
    
    // Clean up the title by removing " | WiFi.Report" or similar suffixes
    let title = titleMatch ? titleMatch[1].split('|')[0].trim() : filename.replace('.html', '');
    let description = descMatch ? descMatch[1] : '';
    let keywords = keywordsMatch ? keywordsMatch[1].split(',').map(k => k.trim()) : [];
    
    return { title, description, keywords };
}

// Generate tags from keywords
function generateTags(keywords) {
    // Take first 2-3 most relevant keywords as tags
    return keywords.slice(0, 2).map(kw => {
        // Capitalize first letter of each word
        return kw.split(' ')
            .map(word => word.charAt(0).toUpperCase() + word.slice(1))
            .join(' ');
    });
}

// Estimate reading time based on content length
function estimateReadingTime(htmlContent) {
    // Remove HTML tags and count words
    const text = htmlContent.replace(/<[^>]*>/g, ' ');
    const words = text.split(/\s+/).filter(w => w.length > 0).length;
    // Average reading speed: 200 words per minute
    const minutes = Math.ceil(words / 200);
    return `${minutes} min read`;
}

// Get file creation/modification date
function getFileDate(filepath) {
    const stats = fs.statSync(filepath);
    const date = stats.mtime;
    return date.toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' });
}

// Main function to generate blog list
function generateBlogList() {
    const blogDir = path.join(__dirname, 'blog');
    
    if (!fs.existsSync(blogDir)) {
        console.error('Blog directory not found!');
        return;
    }
    
    const files = fs.readdirSync(blogDir)
        .filter(file => file.endsWith('.html'));
    
    const blogPosts = files.map(file => {
        const filepath = path.join(blogDir, file);
        const htmlContent = fs.readFileSync(filepath, 'utf8');
        
        const { title, description, keywords } = extractMetadata(htmlContent, file);
        const tags = generateTags(keywords);
        const readingTime = estimateReadingTime(htmlContent);
        const date = getFileDate(filepath);
        const icon = blogConfig.icons[file] || blogConfig.defaultIcon;
        
        return {
            filename: file,
            url: `blog/${file}`,
            title,
            description,
            tags,
            readingTime,
            date,
            icon,
            dateObj: new Date(date) // For sorting
        };
    });
    
    // Sort by date (newest first)
    blogPosts.sort((a, b) => b.dateObj - a.dateObj);
    
    // Remove dateObj before saving (used only for sorting)
    blogPosts.forEach(post => delete post.dateObj);
    
    // Save to JSON file
    const outputPath = path.join(__dirname, 'blog-posts.json');
    fs.writeFileSync(outputPath, JSON.stringify(blogPosts, null, 2));
    
    console.log(`✅ Generated blog list with ${blogPosts.length} posts`);
    console.log(`📝 Saved to: ${outputPath}`);
}

// Run the script
generateBlogList();
